### ISSUE TITLE
Question/Request - Is there a way to force StaxRip to retain Template settings?

### ISSUE THREAD

**Issue by noblemd00** [staxrip/staxrip#1431]
v.2.41.7  NVEncC H265

Apparently, based on the original encoding, StaxRip arbitrarily changes user Template settings when loading source files - Can this be prevented?

i.e. VUI settings -Colormatrix, Colorprim, Transfer, and etc. are consistently changed.

If there is no option to prevent this from happening - request an option to force retention of template settings. Or, at least a notification and permission to change user set template options.

1. The user should be able to set options to encode files as the user needs without them being changed.

2. In order to play DolbyVision files on some TVs certain color settings must be used - StaxRip usually changes these user template settings forcing the user to manually reenter the setting(s) after source loading. Note: some files retain the proper settings.

i.e. Template VUI Transfer setting = BT 2020-12/10 or Undefined - After DolbyVision source load, VUI Transfer must be manually re-set to Undefined or BT 2020-10/12 for some TVs to recognize the video as a DolbyVision encode - StaxRip always changes a DolbyVision encode VUI Transfer to SMPTE 2084 which will not play as DolbyVision on some TVs. 

3 And, cannot Batch process files with user required settings - settings are changed by StaxRip!

**Comment by Dendraspis** [comment]
A very long question/request for a very short answer:  `Project » Options » Video » Import VUI metadata`  😁 

Or as a picture:
<img width=800 src=https://github.com/user-attachments/assets/aabe87d2-954d-404a-9678-7e36751714ca>

The red marked settings are especially used to alter the template settings.
Note, that the Dolby Vision metadata as well as the RPU file path are also set via `Import VUI metadata`. So in case your mentioned workflow you would need to add it by hand.

> If there is no option to prevent this from happening

😅 

> After DolbyVision source load, VUI Transfer must be manually re-set to Undefined or BT 2020-10/12 for some TVs to recognize the video as a DolbyVision encode

Which Dolby Vision profile is it?

**Comment by noblemd00** [comment]
> After DolbyVision source load, VUI Transfer must be manually re-set to Undefined or BT 2020-10/12 for some TVs to recognize the video as a DolbyVision encode
> 
> Which Dolby Vision profile is it?

Dolby Vision 8.1

**Comment by noblemd00** [comment]
Thanks for the reply! - I've noticed that VUI Import setting several times but the purpose never clicked. Doh!

**Comment by Dendraspis** [comment]
In 99.9% of all cases, the settings made by `Import VUI metadata` will get a working and proper result. I don't know why your TV should have any issues with it, but in general it is what you should use.
The only exception should be Profile 5.

> I've noticed that VUI Import setting several times but the purpose never clicked. Doh!

Happens sometimes.  😁 

So I guess we can close this one?

**Comment by noblemd00** [comment]
If Import VUI metadata is unchecked, Dolby Vision RPU and HDR file paths are not loaded in VUI Options requiring editing after source load (and preventing batch processing). 

When loading Dolby Vision or HDR encoded files the RPU/Json file/path-name should be loaded even when Import VUI metadata is unchecked. OR, additional options should be available to choose what to load.

**Comment by Dendraspis** [comment]
> If Import VUI metadata is unchecked, Dolby Vision RPU and HDR file paths are not loaded in VUI Options requiring editing after source load (and preventing batch processing).

Yes, that's what I already mentioned in my first post.  😅    

Those files are part of VUI and I don't think I will change the behavior.
For the paths you can use macros to point to those files, something like:
`%temp_dir%\HDRDVmetadata.rpu`

**Comment by noblemd00** [comment]
I've added an Event Command that seems to work...

My templates name (for files needing an RPU) contains either "DV" (Dolby Vision only) or "HDV" (DV+HDR) - the event command checks for "DV" in the template name then executes the command line --dolby-vision-rpu  "%temp_dir%\HDRDVmetadata.rpu" if present. 

I also plan to create an Event Command to add the HDR10 info file automatically. Any suggestions for that would appreciated.


<img width="658" alt="240723 101911" src="https://github.com/user-attachments/assets/0c4ac264-438a-407a-9699-bf03ed8bba8c">

**Comment by Dendraspis** [comment]
It works, but:   ***Why so complicated ????***

Just use what I have mentioned above and you're fine:
<img width=600 src=https://github.com/user-attachments/assets/5369eacb-a56b-4073-8da2-d085bda0a5ae>

**Comment by Dendraspis** [comment]
Does it work? Can we close this now?

**Comment by noblemd00** [comment]
No, does not work - does not load other needed settings (such as the Master Display entry).

Selected OPTION/VIDEO/Import VUI metadata and...

I've changed the Event Command as below (to change Transfer to a proper setting - for me - still experimenting but it seems to work):

<img width="477" alt="240726 093501" src="https://github.com/user-attachments/assets/03b0821d-15d6-4c21-8218-aba8c66d46fd">

**Comment by noblemd00** [comment]
Additional-

There should be a way (multiple CheckBoxes) to allow the user to prevent the program from changing individual template VUI settings other than the LOAD EVERYTHING or LOAD NOTHING choice.

And, with the added Dolby Vision and HDR(10+) there needs to be Event Command Options/Macros relating to Dolby Vision and HDR(10+) settings.

**Comment by Dendraspis** [comment]
To be fair:  You are the first one with such an issue and I still don't understand why you have to do this....

Basically you use `Import VUI metadata` to transfer all related settings to the encoder and in general this is all you need and want. In your case you can overwrite some of those settings via events.

> And, with the added Dolby Vision and HDR(10+) capabilities there needs to be Event Command Options/Macros relating to Dolby Vision and HDR(10+) settings.

`ImportVideEncoderCommandLine` covers all parameters for the encoders. Sounds like you request taking two or more parameters out of it to make them settable directly, also handling issues when both are set or the different parameters across the encoders. I don't think this is the right way....
When it comes to macros I totally agree with you and I will implement them:  `%hdr10plus_path%` and  `%hdrdv_path%`  I guess the path is enough for every purpose.

> One other thing - Event Commands are still not working when Batch processing. (not sure if you addressed this problem earlier).

I will have a look....

**Comment by noblemd00** [comment]
> When it comes to macros I totally agree with you and I will implement them: %hdr10plus_path% and %hdrdv_path%. I guess the path is enough for every purpose.

Will that allow a differentiation between - Being able to select - HDR10PlusMetadata.json file if available or HDRDVmetadata_L5.json if not - i.e. How will StaxRip know that it has to use the HDRDVmetadata_L5.json file because the HDR10PlusMetadata.json file does not exist? And if neither exist process as Dolby Vision only (Profile 5 - this part not required but might be useful to some USERS ).

<img width="403" alt="240729 143907" src="https://github.com/user-attachments/assets/912c36c0-c30f-4f69-80ef-031f5a1d2a6f">

<img width="153" alt="240729 145343" src="https://github.com/user-attachments/assets/61274da2-4d16-423f-9fcc-5d7b7ae4f8aa">

**Comment by noblemd00** [comment]
One other thing while I'm asking questions - 

It does not matter how I process Dolby Vision videos (different conversion modes, different VUI Dolby Vision Profile setting, and slightly different Color Matrix etc. settings.

I always get finished files that MediaInfo reports:
HDR format : Dolby Vision, Version 1.0, Profile 5, dvhe.05.06, BL+RPU, no metadata compression / SMPTE ST 2086 / SMPTE ST 2086

Am I doing something wrong - I've tried various settings.

**Comment by noblemd00** [comment]
Batch file processing - This event inserts height in the Target name (1080p, 720p, or whatever) - works for single processed file but NOT Batch processed files. I don't think any Events are working during Batch proc. Have not checked all.

> Which event(s) do you mean in particular?

<img width="766" alt="240729 194142" src="https://github.com/user-attachments/assets/8fb0b9fe-569f-4f5f-b979-bcec94a1018c">

**Comment by Dendraspis** [comment]
The Command is not really interesting for me at the moment, I would like to know which `Event` it is that you have chosen in the upper right corner, so I can check exactly where it gets called. 

> Will that allow a differentiation between [..]

`%hdr10plus_path%` is for HDR10+ and `%hdrdv_path%` is for Dolby Vision files. Because you set them before encoding, StaxRip will process the paths as usual.

> I always get finished files that MediaInfo reports [..]

Don't know if you take care, but you also have to set the right `Dolby Vision Profile`:
<img width=500 src=https://github.com/user-attachments/assets/0a33fcfe-95f9-4447-80a5-8e163640a29e>
<img width=500 src=https://github.com/user-attachments/assets/2601a28d-74e7-428a-9c4b-ef3c221051d2>

Profile 5 means, that your video can only be played on players, that can play Dolby Vision - there is no compatibility for HDR10. Therefore the default is Profile 8.1. Like I said, the default values result in a proper Profile 8.1 result:
`Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, no metadata compression, HDR10 compatible / SMPTE ST 2086, Version HDR10, HDR10 compatible`

Maybe you should try a *normal* file processing and not a batch file encoding, just to make sure, that it is not Batch encoding itself or a wrong template or so.

**Comment by noblemd00** [comment]
BTW I just tried to load a Dolby Vision Profile 5 and the RPU path/file was not loaded into VUI (I think I reported this problem a long time ago).

**Comment by noblemd00** [comment]
Another FYI:
In order to properly re-encode Dolby Vision 8 and HDR10+ files 

0: Parses RPU, rewrites it untouched

must be selected as the RPU Conversion Mode

MediaInfo with RPU Conversion Mode:  2: Convert RPU to be 8.1 compatible
HDR format : Dolby Vision, Version 1.0, Profile 8, dvhe.08.06, BL+RPU, no metadata compression / SMPTE ST 2094 App 4

MediaInfo with RPU Conversion Mode:   0: Parses RPU, rewrites it untouched
HDR format : Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, no metadata compression, HDR10 compatible / SMPTE ST 2094 App 4, Version HDR10+ Profile B, HDR10+ Profile B compatible

**Comment by Dendraspis** [comment]
Okay, so this issue seems to be fixed thanks to an up to date KODI version.  🥳  Glad we can at least close that chapter...

> NVEncC - H.265

Can't confirm that, but it depends which meta dates the source file provides. Profile 5 sources often leak substantial HDR metadata , so the VUI Import is not called, which means that the metadata files are also not set. `v2.42` will try to close the gap and set the metadata files regardless the other information.

> In order to properly re-encode Dolby Vision 8 / HDR10+ combo files

At first place that depends on your source file and what you want your result to be.
Dolby Vision wise both encodes should be the same. In your case you also seem to have HDR10+ metadata, which can make the difference. In that case it would be interesting to see the media info of the source file.

**Comment by noblemd00** [comment]
Re: Another FYI (DV/HDR10+)

Additional settings (may be needed - haven't tested every option)
<img width="338" alt="240730 202947" src="https://github.com/user-attachments/assets/b5333c7e-fcf4-4450-8248-8c5c01046d2c">

Yields:
Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, no metadata compression, HDR10 compatible / SMPTE ST 2094 App 4, Version HDR10+ Profile B, HDR10+ Profile B compatible

Not sure if these can be altered, but this was gleaned from previous HDR10+ experiments and they worked whereas other settings did not.

**Comment by noblemd00** [comment]
Original file from DV/HDR10+
[A.Quiet.Place.Day.One.2024.HDV.mkv.MediaInfo.txt](https://github.com/user-attachments/files/16434845/A.Quiet.Place.Day.One.2024.HDV.mkv.MediaInfo.txt)

**Comment by Dendraspis** [comment]
When both meta files are set basically there should be no difference, especially because the Dolby Vision part of the media info is the same, but I have to find such a sample file to verify that.

You can use the `RPU Conversion Mode 0` for such cases as the default setting is just a default.
If there is a significant difference, I might add an *Automatic* mode, but not sure about that.


> Additional settings (may be needed - haven't tested every option)

Generally no VUI setting changes should be necessary as they are imported from the source. Changing them can interfere with playback compatibilities.

**Comment by noblemd00** [comment]
I've refined the previous DV/HDR10+ VUI settings - testing several combinations - to get the proper HDR Format (per MediaInfo as below):

Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, no metadata compression, HDR10 compatible / SMPTE ST 2094 App 4, Version HDR10+ Profile B, HDR10+ Profile B compatible

and get proper Dolby Vision playback while also getting the best HDR10+/HDR playback.

These settings seem to work best (RPU Conversion Mode 0):
  
<img width="315" alt="240801 145708" src="https://github.com/user-attachments/assets/bcc25c53-39de-42fa-b918-898e27f5e1ca">

**Comment by Dendraspis** [comment]
Looks exactly like intended, so the only difference seems to be the *RPU Conversion Mode*.  🤔 

I will think about adding an *Automatic* mode to catch that case, but I don't think it's worth it, especially because both results should be the same, at least valid.

**Comment by noblemd00** [comment]
This is the matrix I have come up with after much experimentation...

<img width="616" alt="240801 182456" src="https://github.com/user-attachments/assets/79a41f29-ef6d-410a-ad75-0ba4856f212b">

Note: DV+HLG re-encoded from DV/HDR(10+) videos.
 All yield proper MediaInfo HDR Format entries.

**Comment by noblemd00** [comment]
Update to the MATRIX

<img width="616" alt="240802 162700" src="https://github.com/user-attachments/assets/1c86330d-f023-435e-9e63-436b72477f80">

**Comment by noblemd00** [comment]
Am I correct that the RPU Conversion Mode needs to be set before source load - In other words, the RPU conversion "converts" as the source file is loaded into StaxRip or does the RPU conversion take place during processing?

**Comment by Dendraspis** [comment]
Currently:  yes.

I'm thinking about an automatic mode or adding a new mode for specific cases, but it's not that easy to take every aspect into account. For the moment it seems to be enough to add a new mode, that uses RPU conversion mode 2 generally, but mode 0 for DV Profile 5.

**Comment by Dendraspis** [comment]
Let's keep it open till I have found a solution...

**Comment by noblemd00** [comment]
Minor update to the matrix re. re-encoding HDR10+ only videos - Depending on how the HDR10+ video was originally encoded, The  ATC SEI either has to be "Auto" or "SMPTE 2084" check MediaInfo after re-encoded for proper configuration.

<img width="656" alt="_  Re-encode VUI Parameters" src="https://github.com/user-attachments/assets/32b8b8c4-897c-4370-91cb-b24712608f7e">

**Comment by Dendraspis** [comment]
Thanks for the overview, I will have a look and see what I can improve.

Btw:  Did you already join the Discord server?

**Comment by Dendraspis** [comment]
I guess that means *no*.  😁 

Just to make sure:
`DV+HLG` uses Profile `8.4`, so the conversion mode should be `4` instead of `2`. Do you agree?

**Comment by noblemd00** [comment]
Updated matrix (above) per suggestion - DV+HLG uses Profile 8.4...

**Comment by noblemd00** [comment]
Question about the new macros %hdr10plus_path%, %hdrdv_path%

How can a determination be made (automatically) to use the proper HDRDVmetadata_L5.json, HDRDVmetadata_Config.json or HDR10PlusMetadata.json entry in the HDR10 info file inputbox?

Event Command: ImportVideoEncoderCommandLine (After Source Opened)
Command Line = --dhdr10-info "%hdr10plus_path%" 

If the video file is HDR10+ the entry is correct however if the file is DV/HDR or DV/HDR10 (not +) the entry is blank.

**Comment by Dendraspis** [comment]
Welcome back.  😁 

> How can a determination be made (automatically) to use the proper HDRDVmetadata_L5.json, HDRDVmetadata_Config.json or HDR10PlusMetadata.json entry in the HDR10 info file inputbox?

The macro will point to the right file. Also `HDRDVmetadata_L5.json` and `HDRDVmetadata_Config.json` are files to help the Dolby Vision feature (as the name suggests), they are not relevant for the encoder.

>  if the file is DV/HDR or DV/HDR10 (not +)

Which is the same by definition.

> If the video file is HDR10+ the entry is correct however if the file is DV/HDR or DV/HDR10 (not +) the entry is blank.

Of course the macro cannot point at a file that does not exist, because it could not be extracted.


Meanwhile since `v2.42` the Dolby Vision Profile 5 metadata is also set via `Import VUI metadata`, so that issue should be gone. And for each situation or configuration everything can be set up by choosing the right options. So from my side this issue seems to be done. There should be no reason to bypass this feature and when you think you have to do it, you have to do all steps on your own.

Nevertheless I think I will make some more improvements/changes regarding the Dolby Vision handling in `v2.43`.

**Comment by noblemd00** [comment]
Disabled all Event commands that set HDR Info file... 

HDR format: Dolby Vision, Version 1.0, Profile 7.6, dvhe.07.06, BL+EL+RPU, no metadata compression, Blu-ray compatible / SMPTE ST 2086, Version HDR10, HDR10 compatible

HDRDVmetadata_L5.json file is present in Temp dir.

and after loading source the HDR10 Info File was blank (HDRDVmetadata.rpu entry ok).

Same thing happened with...
HDR format: Dolby Vision, Version 1.0, Profile 8.1, dvhe.08.06, BL+RPU, no metadata compression, HDR10 compatible / SMPTE ST 2086, Version HDR10, HDR10 compatible

`The macro will point to the right file. Also HDRDVmetadata_L5.json and HDRDVmetadata_Config.json are files to help the Dolby Vision feature (as the name suggests), they are not relevant for the encoder.`

I was under the impression that the HDR10 Info file (HDRDVmetadata_L5.json, HDRDVmetadata_Config.json) was required for proper compatible playback on HDR10 (not DV) TVs? 

But, after testing, it appears they are not (re-encoded files with blank HDR10 Info File entry playback as HDR - see settings below). 

Also for proper compatible playback on **HDR(10+) only TV** the following settings are required: 
Tested 'Auto' for ATC SEI and the result was no HDR playback.
Colorrange 'Full' resulted in the picture being to bright.

<img width="361" alt="240911 095408" src="https://github.com/user-attachments/assets/23104e01-e749-48a5-86b8-79a44bd0c823">

**Comment by Dendraspis** [comment]
> and after loading source the HDR10 Info File was blank (HDRDVmetadata.rpu entry ok).

That's good.

> I was under the impression that the HDR10 Info file (HDRDVmetadata_L5.json, HDRDVmetadata_Config.json) was required for proper compatible playback on HDR10 (not DV) TVs?

`HDR10 Info file` is a for `HDR10+` JSON files.

> Also for proper compatible playback on HDR(10+) only TV the following settings are required:

All TVs should be able to play `HDR10` which is available in both of your sources. But there is nothing like `HDR10+ only`.

> Also, with the settings below playback on Dolby Vision TV is proper.

Looks like the default settings for me.

**Comment by noblemd00** [comment]
`But there is nothing like HDR10+ only`

I'm trying to figure out what you mean by this - HDR10+ only encoded video files (without Dolby Vision encoding) definitely exist!

I have video files that are MediaInfo listed as...
HDR format: SMPTE ST 2094 App 4, Version HDR10+ Profile B, HDR10+ Profile B compatible 
and
HDR format: SMPTE ST 2094 App 4, HDR10+ Profile B compatible

**Comment by Dendraspis** [comment]
> I'm trying to figure out what you mean by this - HDR10+ only encoded video files (without Dolby Vision encoding) definitely exist!

`HDR10` is not `HDR10+` !!

`HDR10+` builds up on top of `HDR10`, so every `HDR10+` video has a `HDR10` fallback. `Dolby Vision` is similar, but can be used without `HDR10` layer by using `Profile 5`.

**Comment by noblemd00** [comment]
I can add this as a bug if it is a bug or leave it here if I misunderstand the purpose of Default Target Name-Macros.

Using MACROS in Options-Path-Default Target Name does not seem to do anything (am I understanding the purpose?)

Default Target Name entry: %source_file%%target_height%p%encoder_ext%

Source name: "0107.VII.mkv" with target cropped to 960 and encoded as h265

Should yield a finished (Target) file name of:  0107.VII.960p.h265.mkv

but the target name ends up as: "0107.VII.mkv" (same as source name nothing changed/added with the Default Target Name entry not doing anything).

**Comment by noblemd00** [comment]
I made this request a long time ago... (and I just tried to find an encoded file which did not end up where I wanted it - the file was buried in the C: drive USER folder due to a very minor mis-type.) 

Yes this info is sometimes available - buried in the Log in FFMeg/MediaInfo listings but it is not prominently displayed - not easily seen - usually requiring a keyword text search to find (and sometimes finished file location info is not in the log).

The finished VIDEO FILE LOCATION and FILE NAME should be the final entry in the log (even better yet would be the FIRST entry in the LOG - not sure if this is possible).

Since the purpose of StaxRip is to create a new video file - shouldn't the finished file location information be prominently included in the LOG?

The (easy to find) entry in the LOG should be (something similar to):

ENCODED-TARGET FILE PATH/NAME : 
X:\Video Folder\VideoFileName.ext

After all, creating the re-encoded file IS THE PURPOSE of StaxRip!
Where is the StaxRip generated file - What is the StaxRip generated file name?
